### DISCUSSION TITLE
无法从 nonebot2 中的 plugin 包导入 on 模块

### DISCUSSION THREAD

**Discussion by A-kirami** [nonebot/discussions#75] [问答] [answered]
因为 nonebot2 在 nonebot.plugin 的第 16 行从 on 模块中导入了重名的 on 函数 ：`from .on import on as on`，所以现在使用`from nonebot.plugin import on`导入的是 on 函数，而不是 on 模块

现在需要覆盖 on 模块内的部分函数（诸如`on_command`、`on_keyword`之类），但由于上述问题，无法成功覆盖

所以想问一下如何操作才能正确覆盖原函数，十分感谢

**Comment by yanyongyu** [comment, marked as answer]
```python
import nonebot.plugin.on as on
```

**Reply by A-kirami** [reply]
尝试之后发现并没有成功覆盖，代码如下：

```python
import nonebot.plugin.on as on

def on_message(**kwargs):
    print("new_on_message")

on.on_message = on_message
```

```python
import inspect
from nonebot.plugin import on_message

hello = on_message()

print(inspect.getsource(on_message))
```

打印出来的还是 nb2 原本的 on_message 代码

**Reply by yanyongyu** [reply]
> on.on_message = on_message
> from nonebot.plugin import on_message

你从两个地方import的on_message自然是不一样的

你覆盖的是on里的，并不会影响到nonebot.plugin里的

**Reply by A-kirami** [reply]
🥳明白了，谢谢大佬的回答
